src/Database/Statement.php

```diff
@@ -0,0 +1,399 @@
+        if ($type === null) {
+            $type = 'string';
+        }
+        if (!is_int($type)) {
+            [$value, $type] = $this->cast($value, $type);
+        }
+        $this->performBind($column, $value, $type);
+    }
+
+    /**
+     * Perform bind to statement.
+     *
+     * @param string|int $column name or param position to be bound
+     * @param mixed $value The value to bind to variable in query
+     * @param int $type PDO::PARAM_* constant
+     * @return void
+     */
+    protected function performBind(string|int $column, mixed $value, int $type): void
+    {
+        $this->statement->bindValue($column, $value, $type);
+    }
+
+    /**
+     * Executes the statement by sending the SQL query to the database. It can optionally
+     * take an array or arguments to be bound to the query variables. Please note
+     * that binding parameters from this method will not perform any custom type conversion
+     * as it would normally happen when calling `bindValue`.
+     *
+     * @param array|null $params list of values to be bound to query
+     * @return bool true on success, false otherwise
+     */
+    public function execute(?array $params = null): bool
+    {
+        $this->rows = [];
+        $this->rowsFetched = false;
+        $this->executed = true;
+
+        $result = $this->statement->execute($params);
+        if ($result && $this->logger) {
+            //$this->logger->debug();
+        }
+
+        return $result;
+    }
+
+    /**
+     * Closes a cursor in the database, freeing up any resources and memory
+     * allocated to it. In most cases you don't need to call this method, as it is
+     * automatically called after fetching all results from the result set.
+     *
+     * @return void
+     */
+    public function closeCursor(): void
+    {
+        $this->statement->closeCursor();
+    }
+
+    /**
+     * Returns the next row for the result set after executing this statement.
+     * Rows can be fetched to contain columns as names or positions. If no
+     * rows are left in result set, this method will return false.
+     *
+     * ### Example:
+     *
+     * ```
+     * $statement = $connection->prepare('SELECT id, title from articles');
+     * $statement->execute();
+     * print_r($statement->fetch('assoc')); // will show ['id' => 1, 'title' => 'a title']
+     * ```
+     *
+     * @param int $mode Fetch mode. PDO::FETCH_* constant or name.
+     * @return mixed
+     */
+    public function fetch(string|int $mode = 'num'): mixed
+    {
+        $mode = static::MODES[$mode] ?? $mode;
+        if (is_string($mode)) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid named fetch mode `%s` requested. Must be one of %s.',
+                $mode,
+                array_keys(static::MODES)
+            ));
+        }
+
+        if ($this->bufferRows && $this->rowsFetched) {
+            // Can we validate the fetch mode matches? It might be row-specific.
```

Comment:
I am wondering if we should just drop the modes and always use `FETCH_ASSOC`. Since we no longer decorate `PDOStatement` we don't need to replicate it's API.

Reply:
Allowing multiple modifies combined with buffering has caused issues like #14676.

Reply:
That is due to bugs like this: https://github.com/cakephp/cakephp/issues/16214.

We should not re-create this implementation.

Reply:
If we go away from the `\PDOStatement` wrapper approach we could make the fetch mode a property like buffering that can't be mutated once rows have been loaded.

Another option would be to 'remember' what the first mode is and then raise an error if the mode changes.

Reply:
The issue we're running into is Association code doing a fetchAll('assoc') on every query so we can't really support arbitrary fetch modes for ORM select queries. 

We were discussing how to limit this. I had the idea of making buffering internal and only through ORM select queries. You couldn't `execute()` them manually.

Other queries would get a statement that directly wraps PDOStatement with all fetches available.

Reply:
Could ORM\Query also implement fetchAll() and remove the other fetch modes?

Not having other fetch modes is likely ok for ORM queries but removing them from Database\Query will cause problems for at least a few applications. We will also cause bigger breaks if we stray from the PDOStatement interface though.

I am wondering if having Orm\Query wrap a Database\Query insteadnof subclassing would help solve other design issues we are facing 🤔

Reply:
We could move the buffering out of Statement  entirely and build the result set from the association results merged with the statement fetch.

Reply:
> We could move the buffering out of Statement entirely and build the result set from the association results merged with the statement fetch.

This could work well too. Another option could also be behavior that is moved up into ORM\Query or new ORM\Statement :thinking: